File: lib/go/yock/json.go
// Copyright 2023 The Yock Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package liby

import (
	"encoding/json"
	"errors"

	yocki "github.com/ansurfen/yock/interface"
	lua "github.com/yuin/gopher-lua"
)

func LoadJSON(yocks yocki.YockScheduler) {
	lib := yocks.CreateLib("json")
	lib.SetFunctions(map[string]lua.LGFunction{
		"encode": apiEncode,
		"decode": apiDecode,
	})
}

// power by https://github.com/layeh/gopher-json
// The following functions are exposed by the library:
//  decode(string): Decodes a JSON string. Returns nil and an error string if
//                  the string could not be decoded.
//  encode(value):  Encodes a value into a JSON string. Returns nil and an error
//                  string if the value could not be encoded.
//
// The following types are supported:
//
//  Lua      | JSON
//  ---------+-----
//  nil      | null
//  number   | number
//  string   | string
//  table    | object: when table is non-empty and has only string keys
//           | array:  when table is empty, or has only sequential numeric keys
//           |         starting from 1
//
// Attempting to encode any other Lua type will result in an error.
//
// Example
//
// Below is an example usage of the library:
//  import (
//      luajson "layeh.com/gopher-json"
//  )
//
//  L := lua.NewState()
//  luajson.Preload(L)

// @param str string
//
// @return table
func apiDecode(L *lua.LState) int {
	str := L.CheckString(1)

	value, err := Decode(L, []byte(str))
	if err != nil {
		L.Push(lua.LNil)
		L.Push(lua.LString(err.Error()))
		return 2
	}
	L.Push(value)
	return 1
}

// @param value any
//
// @return string
func apiEncode(L *lua.LState) int {
	value := L.CheckAny(1)

	data, err := Encode(value)
	if err != nil {
		L.Push(lua.LNil)
		L.Push(lua.LString(err.Error()))
		return 2
	}
	if L.GetTop() == 3 {
		jsonString := string(data)
		var jsonData interface{}
		err := json.Unmarshal([]byte(jsonString), &jsonData)
		if err != nil {
			L.Push(lua.LNil)
			L.Push(lua.LString(err.Error()))
			return 2
		}

		prettyJSON, err := json.MarshalIndent(jsonData, "", "    ")
		if err != nil {
			L.Push(lua.LNil)
			L.Push(lua.LString(err.Error()))
			return 2
		}
		L.Push(lua.LString(prettyJSON))
		return 1
	}
	L.Push(lua.LString(string(data)))
	return 1
}

var (
	errNested      = errors.New("cannot encode recursively nested tables to JSON")
	errSparseArray = errors.New("cannot encode sparse array")
	errInvalidKeys = errors.New("cannot encode mixed or invalid key types")
)

type invalidTypeError lua.LValueType

func (i invalidTypeError) Error() string {
	return `cannot encode ` + lua.LValueType(i).String() + ` to JSON`
}

// Encode returns the JSON encoding of value.
func Encode(value lua.LValue) ([]byte, error) {
	return json.Marshal(jsonValue{
		LValue:  value,
		visited: make(map[*lua.LTable]bool),
	})
}

type jsonValue struct {
	lua.LValue
	visited map[*lua.LTable]bool
}

func (j jsonValue) MarshalJSON() (data []byte, err error) {
	switch converted := j.LValue.(type) {
	case lua.LBool:
		data, err = json.Marshal(bool(converted))
	case lua.LNumber:
		data, err = json.Marshal(float64(converted))
	case *lua.LNilType:
		data = []byte(`null`)
	case lua.LString:
		data, err = json.Marshal(string(converted))
	case *lua.LTable:
		if j.visited[converted] {
			return nil, errNested
		}
		j.visited[converted] = true

		key, value := converted.Next(lua.LNil)

		switch key.Type() {
		case lua.LTNil: // empty table
			data = []byte(`[]`)
		case lua.LTNumber:
			arr := make([]jsonValue, 0, converted.Len())
			expectedKey := lua.LNumber(1)
			for key != lua.LNil {
				if key.Type() != lua.LTNumber {
					err = errInvalidKeys
					return
				}
				if expectedKey != key {
					err = errSparseArray
					return
				}
				arr = append(arr, jsonValue{value, j.visited})
				expectedKey++
				key, value = converted.Next(key)
			}
			data, err = json.Marshal(arr)
		case lua.LTString:
			obj := make(map[string]jsonValue)
			for key != lua.LNil {
				if key.Type() != lua.LTString {
					err = errInvalidKeys
					return
				}
				obj[key.String()] = jsonValue{value, j.visited}
				key, value = converted.Next(key)
			}
			data, err = json.Marshal(obj)
		default:
			err = errInvalidKeys
		}
	default:
		err = invalidTypeError(j.LValue.Type())
	}
	return
}

// Decode converts the JSON encoded data to Lua values.
func Decode(L *lua.LState, data []byte) (lua.LValue, error) {
	var value interface{}
	err := json.Unmarshal(data, &value)
	if err != nil {
		return nil, err
	}
	return DecodeValue(L, value), nil
}

// DecodeValue converts the value to a Lua value.
//
// This function only converts values that the encoding/json package decodes to.
// All other values will return lua.LNil.
func DecodeValue(L *lua.LState, value interface{}) lua.LValue {
	switch converted := value.(type) {
	case bool:
		return lua.LBool(converted)
	case float64:
		return lua.LNumber(converted)
	case string:
		return lua.LString(converted)
	case json.Number:
		return lua.LString(converted)
	case []interface{}:
		arr := L.CreateTable(len(converted), 0)
		for _, item := range converted {
			arr.Append(DecodeValue(L, item))
		}
		return arr
	case map[string]interface{}:
		tbl := L.CreateTable(0, len(converted))
		for key, item := range converted {
			tbl.RawSetH(lua.LString(key), DecodeValue(L, item))
		}
		return tbl
	case nil:
		return lua.LNil
	}

	return lua.LNil
}

File: lib/yock/json.lua
--  Copyright 2023 The Yock Authors. All rights reserved.
--  Use of this source code is governed by a MIT-style
--  license that can be found in the LICENSE file.

---@diagnostic disable: duplicate-doc-field
---@diagnostic disable: duplicate-set-field

---@class jsonobject
---@field buf table
---@field file string
jsonobject = {}

---@param file string
---@param str? string
---@return jsonobject
function json.create(file, str)
    local obj = {
        file = file
    }
    if not find(file) then
        write(file, str or "{}")
    end
    local stream, err = cat(file)
    yassert(err)
    obj.buf = json.decode(stream)
    setmetatable(obj, { __index = jsonobject })
    return obj
end

---@param file string
---@return jsonobject
function json.open(file)
    local obj = {
        file = file
    }
    local stream, err = cat(file)
    yassert(err)
    obj.buf = json.decode(stream)
    setmetatable(obj, { __index = jsonobject })
    return obj
end

---@param k string
---@return boolean
function jsonobject:getbool(k)
    local v = self:get(k)
    if type(v) == "string" then
        return v
    end
    return false
end

---@param k string
---@return number
function jsonobject:getnumber(k)
    local v = self:get(k)
    if type(v) == "number" then
        return v
    end
    return 0
end

---@param k string
---@return string
function jsonobject:getstr(k)
    local v = self:get(k)
    if type(v) == "string" then
        return v
    end
    return ""
end

---@param k string
---@return table
function jsonobject:gettable(k)
    local v = self:get(k)
    if type(v) == "table" then
        return v
    end
    return {}
end

---@param k string
---@return any
function jsonobject:rawget(k)
    return self.buf[k]
end

---@param k string
---@return any
function jsonobject:get(k)
    local keys = strings.Split(k, ".")
    local x = self.buf
    for _, key in ipairs(keys) do
        if x == nil then
            return nil
        end
        x = x[key]
    end
    return x
end

---@param k string
---@param v any
function jsonobject:rawset(k, v)
    self.buf[k] = v
end

---@param k string
---@param v any
function jsonobject:set(k, v)
    local fields = strings.Split(k, ".")
    local currentTable = self.buf
    for i = 1, #fields - 1 do
        local field = fields[i]
        if not currentTable[field] then
            currentTable[field] = {}
        end
        currentTable = currentTable[field]
    end
    currentTable[fields[#fields]] = v
end

---@return string
function jsonobject:string()
    return json.encode(self.buf)
end

---@param pretty? boolean
function jsonobject:save(pretty)
    if pretty then
        write(self.file, json.encode(self.buf, "", "    "))
        return
    end
    write(self.file, json.encode(self.buf))
end
